include/SFML/System/Span.hpp

```diff
@@ -0,0 +1,43 @@
+////////////////////////////////////////////////////////////
+//
+// SFML - Simple and Fast Multimedia Library
+// Copyright (C) 2007-2022 Laurent Gomila (laurent@sfml-dev.org)
+//
+// This software is provided 'as-is', without any express or implied warranty.
+// In no event will the authors be held liable for any damages arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it freely,
+// subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented;
+//    you must not claim that you wrote the original software.
+//    If you use this software in a product, an acknowledgment
+//    in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such,
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+#ifndef SFML_SYSTEM_SPAN_HPP
+#define SFML_SYSTEM_SPAN_HPP
+
+////////////////////////////////////////////////////////////
+// Include <span> if C++20 or greater is used.
+// Otherwise, use an alternative implementation for older compilers.
+////////////////////////////////////////////////////////////
+#if __cplusplus >= 202002L || (defined(_MSVC_LANG) && _MSVC_LANG >= 202002L)
+
+    #include <span>
+
+#else
+
+    #define TCB_SPAN_NAMESPACE_NAME std
```

Comment:
Isn't it [UB to define custom symbols in `std`](https://en.cppreference.com/w/cpp/language/extending_std)?

Reply:
It would also cause problems if the user in fact uses C++20, with `std::span` available, but defined differently.

Reply:
Right. I guess this could be ok in an application but that is a problem for a library.
So we could have `sf::span` instead for example or `sf::Span` for consistency with SFML's naming conventions.
I wonder if `sf::Span` could be made compatible with `std::span` (the "real" one). That is not absolutely necessary but could be nice.

Reply:
Can we not conditionally compile?
```cpp
#ifdef CPP_AT_LEAST_20
   #include <3rd-party-span.h> // namespace tcb
   namespace sf { 
       using Span = tcb::span;
   }
#else
   #include <span>
   namespace sf { 
       using Span = std::span;
   }
#endif
```

Someone said `__cplusplus` is not reliable on some Visual Studio versions?

Reply:
Maybe use C++20's feature testing macros?

https://en.cppreference.com/w/cpp/feature_test

The macro for `std::span` is `__cpp_lib_span`.

Reply:
That's even better -- assuming the feature macros _are_ properly supported on all major compilers.

Reply:
> Can we not conditionally compile?

I am not sure. Could it create issues if the library is compiled for C++17 and used in a C++20 project?
Maybe the C++ version detection could happen only when the library is compiled?

> Someone said __cplusplus is not reliable on some Visual Studio versions?

It needs (or it needed at some point) a command line option to be reliable. See [this Microsoft blog post](https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/). I did not test it though.

Reply:
> However, I think we're best off doing what SFML 2 does by simply reimplementing library features that aren't yet available. In the same way SFML 2 has the equivalent of C++11's thread, SFML 3 will have the equivalent of C++20's span.

I agree, that seems better. In addition, spans are constructible from anything with a `.data()` and `.size()` methods so that should be compatible with `std::span` without doing anything.